journal.py

```diff
@@ -16,137 +13,99 @@
-SELECT id, title, text, created FROM entries ORDER BY created DESC LIMIT 1
-"""
-
-DB_FILTER = """
-SELECT id, title, text, created FROM entries WHERE id=%s
-"""
-
-INSERT_ENTRY = """
-INSERT INTO entries (title, text, created) VALUES(%s, %s, %s);
-"""
-
-UPDATE_ENTRY = """
-UPDATE entries SET title=%s, text=%s, created=%s WHERE id=%s;
-"""
 
-DELETE_ENTRY = """
-DELETE FROM entries WHERE id=%s;
-"""
+DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))
+Base = declarative_base()
 
-logging.basicConfig()
-log = logging.getLogger(__file__)
 
-
-def connect_db(settings):
-    """Return a connection to the configured database"""
-    return psycopg2.connect(settings['db'])
-
-
-def init_db():
-    """Create database tables defined by DB_SCHEMA
-    """
-    settings = {}
-    settings['db'] = os.environ.get(
-        'DATABASE_URL', 'dbname=learning_journal user=roberthaskell'
+class Entry(Base):
+    __tablename__ = 'entries'
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
+    title = sa.Column(sa.Unicode(127), nullable=False)
+    text = sa.Column(sa.UnicodeText, nullable=False)
+    created = sa.Column(
+        sa.DateTime, nullable=False, default=datetime.today
     )
-    with closing(connect_db(settings)) as db:
-        db.cursor().execute(DB_SCHEMA)
-        db.commit()
-
-
-@subscriber(NewRequest)
-def open_connection(event):
-    request = event.request
-    settings = request.registry.settings
-    request.db = connect_db(settings)
-    request.add_finished_callback(close_connection)
-
-
-def close_connection(request):
-    """close the database connection for this request
-
-    If there has been an error in the processing of the request, abort any
-    open transactions.
-    """
-    db = getattr(request, 'db', None)
-    if db is not None:
-        if request.exception is not None:
-            db.rollback()
-        else:
-            db.commit()
-        request.db.close()
-
-
-def write_entry(request):
-    """write a single entry to the database"""
-    title = request.params['title']
-    text = request.params['text']
-    created = datetime.today()
-    request.db.cursor().execute(INSERT_ENTRY, (title, text, created))
-    # write to a separate file
-    return {}
-
-
-def update(request, identification):
-    title = request.params['title']
-    text = request.params['text']
-    created = datetime.today()
-    request.db.cursor().execute(UPDATE_ENTRY, (title, text, created, identification))
-    return {}
 
+    def __repr__(self):
+        return u"{}: {}".format(self.__class__.__name__, self.title)
+
+    @classmethod
+    def all(cls, session=None):
+        if session is None:
+            session = DBSession
+        return session.query(cls).order_by(cls.created.desc()).all()
+
+    @classmethod
+    def by_id(cls, id, session=None):
+        if session is None:
+            session = DBSession
+        return DBSession.query(cls).filter(cls.id == id).one()
```

Comment:
After you set up `session` to be either the scoped `DBSession` or an unscoped session you create in a unit test, you should use that binding, rather than using `DBSession` here and below.